src/libstd/process.rs

```diff
@@ -0,0 +1,835 @@
+        self.inner.write(buf)
+    }
+}
+
+/// A handle to a child procesess's stdout
+pub struct ChildStdout {
+    inner: FileDesc
+}
+
+impl Reader for ChildStdout {
+    fn read(&mut self, buf: &mut [u8]) -> old_io::IoResult<uint> {
+        self.inner.read(buf)
+    }
+}
+
+/// A handle to a child procesess's stderr
+pub struct ChildStderr {
+    inner: FileDesc
+}
+
+impl Reader for ChildStderr {
+    fn read(&mut self, buf: &mut [u8]) -> old_io::IoResult<uint> {
+        self.inner.read(buf)
+    }
+}
+
+/// The `Command` type acts as a process builder, providing fine-grained control
+/// over how a new process should be spawned. A default configuration can be
+/// generated using `Command::new(program)`, where `program` gives a path to the
+/// program to be executed. Additional builder methods allow the configuration
+/// to be changed (for example, by adding arguments) prior to spawning:
+///
+/// ```
+/// use std::old_io::Command;
+///
+/// let mut process = match Command::new("sh").arg("-c").arg("echo hello").spawn() {
+///   Ok(p) => p,
+///   Err(e) => panic!("failed to execute process: {}", e),
+/// };
+///
+/// let output = process.stdout.as_mut().unwrap().read_to_end();
+/// ```
+#[derive(Clone)]
+pub struct Command {
+    inner: CommandImp,
+
+    // Details explained in the builder methods
+    stdin: Option<Stdio>,
+    stdout: Option<Stdio>,
+    stderr: Option<Stdio>,
+}
+
+impl Command {
+    /// Constructs a new `Command` for launching the program at
+    /// path `program`, with the following default configuration:
+    ///
+    /// * No arguments to the program
+    /// * Inherit the current process's environment
+    /// * Inherit the current process's working directory
+    /// * Inherit stdin/stdout/stderr for `run` or `status`, but create pipes for `output`
+    ///
+    /// Builder methods are provided to change these defaults and
+    /// otherwise configure the process.
+    pub fn new<S: AsOsStr + ?Sized>(program: &S) -> Command {
+        Command {
+            inner: CommandImp::new(program.as_os_str()),
+            stdin: None,
+            stdout: None,
+            stderr: None,
+        }
+    }
+
+    /// Add an argument to pass to the program.
+    pub fn arg<S: AsOsStr + ?Sized>(&mut self, arg: &S) -> &mut Command {
+        self.inner.arg(arg.as_os_str());
+        self
+    }
+
+    /// Add multiple arguments to pass to the program.
+    pub fn args<S: AsOsStr>(&mut self, args: &[S]) -> &mut Command {
+        self.inner.args(args.iter().map(AsOsStr::as_os_str));
+        self
+    }
+
+    /// Inserts or updates an environment variable mapping.
+    ///
+    /// Note that environment variable names are case-insensitive (but case-preserving) on Windows,
+    /// and case-sensitive on all other platforms.
+    pub fn env<S: ?Sized, T: ?Sized>(&mut self, key: &S, val: &T) -> &mut Command where
+        S: AsOsStr, T: AsOsStr
+    {
+        self.inner.init_env_map(env::vars());
```

Comment:
This smells. This initialisation should be encapsulated inside the `inner` implementation.